Simple user search functionality

simong opened the pull request:
Also added user updating logic

nicolaasmatthijs commented on node_modules/gh-rest/lib/rest/user.js:
```diff
@@ -33,6 +33,29 @@ module.exports = function(client) {
     /**
+     * Get all users for an app
+     *
+     * @param  {Number}         [app]                           The application to search users in, defaults to the current application
```
Should be `The id of the app to get the users for. Defaults to the current app` to be consistent with the other rest files.

nicolaasmatthijs commented on node_modules/gh-rest/lib/rest/user.js:
```diff
@@ -33,6 +33,29 @@ module.exports = function(client) {
+     * Get all users for an app
+     *
+     * @param  {Number}         [app]                           The application to search users in, defaults to the current application
+     * @param  {String}         [filter]                        The string to filter users by
```
Should we call this `query`. `filter` suggest that something more advanced is going to be possible

nicolaasmatthijs commented on node_modules/gh-users/lib/api.js:
```diff
@@ -48,6 +58,49 @@ var getUser = module.exports.getUser = function(id, callback) {
 /**
+ * Get a set of users in an application
+ *
+ * @param  {Context}        ctx                 Standard context object containing the current user and the current applicatio
```
typo

nicolaasmatthijs commented on node_modules/gh-users/lib/api.js:
```diff
@@ -48,6 +58,49 @@ var getUser = module.exports.getUser = function(id, callback) {
+ * Get a set of users in an application
+ *
+ * @param  {Context}        ctx                 Standard context object containing the current user and the current applicatio
+ * @param  {Number}         [appId]             The application to search users in, defaults to the current application
```
Same comment as above

nicolaasmatthijs commented on node_modules/gh-users/lib/api.js:
```diff
@@ -48,6 +58,49 @@ var getUser = module.exports.getUser = function(id, callback) {
+ * @param  {Number}         [offset]            The number to start paging from. Defaults to 0
+ * @param  {Function}       callback            Standard callback function
+ * @param  {Object}         callback.err        An error object, if any
+ * @param  {User[]}         callback.users      The users in the application
```
add `The (filtered) users`

nicolaasmatthijs commented on node_modules/gh-users/lib/api.js:
```diff
@@ -48,6 +58,49 @@ var getUser = module.exports.getUser = function(id, callback) {
+ */
+var getUsers = module.exports.getUsers = function(ctx, app, filter, limit, offset, callback) {
+    // Default some parameters
+    app = GrasshopperUtil.getNumberParam(app) || ctx.app.id;
```
Don't we usually do this in the REST layer?

nicolaasmatthijs commented on node_modules/gh-users/lib/api.js:
```diff
@@ -117,6 +170,62 @@ var createUser = module.exports.createUser = function(ctx, appId, userProfile, c
+ * Update a user
+ *
+ * @param  {Context}        ctx                             The current execution context containing the user and application
+ * @param  {Number}         appId                           The id of the user that should be updated
```
Documentation error

nicolaasmatthijs commented on node_modules/gh-users/lib/api.js:
```diff
@@ -117,6 +170,62 @@ var createUser = module.exports.createUser = function(ctx, appId, userProfile, c
+ * @param  {User}           callback.user                   The updated user
+ */
+var updateUser = module.exports.updateUser = function(ctx, userId, update, callback) {
+    var validator = new Validator();
```
userId should be converted to number first

nicolaasmatthijs commented on node_modules/gh-users/lib/api.js:
```diff
@@ -48,6 +58,49 @@ var getUser = module.exports.getUser = function(id, callback) {
+    }
+
+    // Ensure the user is allowed to search/page users in the given application
+    if (app !== ctx.app.id && (!ctx.user || !ctx.user.canAdmin(app))) {
```
Can't we just rely on `isLoggedInUser(ctx, appId)` for this?

nicolaasmatthijs commented on node_modules/gh-users/lib/internal/dao.js:
```diff
@@ -23,6 +23,57 @@ var DB = require('gh-core/lib/db');
+ */
+var init = module.exports.init = function(callback) {
+    // Enable index searching based on trigrams. This assumes that the trigram postgres extension
+    // has been enabled. See http://www.postgresql.org/docs/9.1/static/pgtrgm.html for more information.
```
`@see`

nicolaasmatthijs commented on node_modules/gh-users/lib/internal/dao.js:
```diff
@@ -45,27 +96,72 @@ var getUser = module.exports.getUser = function(id, callback) {
 
 /**
- * Update a user
+ * Get a user
```
Documentation error

nicolaasmatthijs commented on node_modules/gh-users/lib/internal/dao.js:
```diff
@@ -45,27 +96,72 @@ var getUser = module.exports.getUser = function(id, callback) {
- * @param  {Function}       callback                    Standard callback function
- * @param  {Object}         callback.err                An error object, if any
- * @param  {User}           callback.user               The updated user
+ * @param  {Number}         id                              The id of the user to retrieve
```
Documentation errors

nicolaasmatthijs commented on node_modules/gh-users/lib/internal/dao.js:
```diff
@@ -45,27 +96,72 @@ var getUser = module.exports.getUser = function(id, callback) {
- * Get multiple users
+ * Search for users in an application
+ *
+ * @param  {Number}         appId               The application to search users in
```
Same comments as above

nicolaasmatthijs commented on node_modules/gh-users/lib/rest.js:
```diff
@@ -28,14 +28,12 @@ var UsersAPI = require('gh-users');
  * @Method      GET
  * @Path        /users
- * @QueryParam  {number}            [appId]                 The id of the app to retrieve the users for
+ * @QueryParam  {string}            [q]                     The string to filter users by
```
Should be in alphabetical order

nicolaasmatthijs commented on node_modules/gh-users/lib/rest.js:
```diff
@@ -45,13 +43,23 @@ GrassHopper.globalAdminRouter.on('get', '/api/users', function(req, res) {
  * @Server      app
  * @Method      GET
  * @Path        /users
+ * @QueryParam  {string}            [q]                     The string to filter users by
```
Alphabetical order

nicolaasmatthijs commented on node_modules/gh-users/tests/test-search.js:
```diff
@@ -0,0 +1,214 @@
+/**
+ * Copyright (c) 2014 "Fronteer LTD"
+ * Grasshopper Serie Engine
```
This should just be `Grasshopper Event Engine` everywhere

nicolaasmatthijs commented on node_modules/gh-users/tests/test-search.js:
```diff
@@ -0,0 +1,214 @@
+                // user might not get returned. Default the index value to a really high number, so we can more easily
+                // reason about them
+                jackIndex = (jackIndex === -1) ? Number.MAX_VALUE : jackIndex;
+                williamIndex = (williamIndex === -1) ? Number.MAX_VALUE : williamIndex;
```
Why would William and Shirley ever be returned?

simong replied:
Although unlikely, it can happen. The email field is randomised and could potentially have "jac" (or something similar to it) in it.

nicolaasmatthijs replied:
Can we not also set the email on all of the users so we can verify that users are correctly excluded/included?

simong replied:
We might be able to use randomised numbers rather randomised strings. I'll have a look at this

nicolaasmatthijs commented on node_modules/gh-users/tests/test-search.js:
```diff
@@ -0,0 +1,214 @@
+                var williamIndex = _.findIndex(users.results, {'id': william.profile.id});
+                var shirleyIndex = _.findIndex(users.results, {'id': shirley.profile.id});
+
+                // Because the endpoint only returns x number of users, The "William Lawson" and "Shirley Temple"
```
Same here

nicolaasmatthijs commented on node_modules/gh-users/tests/test-search.js:
```diff
@@ -0,0 +1,214 @@
+    it('verify search results can be paged', function(callback) {
+        _setup(function(jack, william, shirley) {
+
+            UsersTestsUtil.getUsers(global.tests.admins.cam2013.client, null, 'i', 2, 0, function(firstPage) {
```
I'd rather test this on a new app (don't rely on previously created users) and verify that the paging stops after the expected number of pages and verify that the results per page are different.

nicolaasmatthijs commented on node_modules/gh-users/tests/test-users.js:
```diff
@@ -0,0 +1,153 @@
+        /**
+         * Test that verifies that a user can be created
+         */
+        it('verify a user can be created', function(callback) {
```
Should probably also have validation and authorization tests for this

nicolaasmatthijs commented on node_modules/gh-users/tests/util.js:
```diff
@@ -37,6 +37,89 @@ var assertGetMe = module.exports.assertGetMe = function(client, callback) {
+var assertUpdateUser = module.exports.assertUpdateUser = function(client, id, update, callback) {
+    client.user.updateUser(id, update, function(err, user) {
+        assert.ok(!err);
+        assert.ok(user);
```
Can we also assert the values on the updated user?

nicolaasmatthijs commented on node_modules/gh-users/tests/util.js:
```diff
@@ -37,6 +37,89 @@ var assertGetMe = module.exports.assertGetMe = function(client, callback) {
+ * Assert that the users for an application can be retrieved
+ *
+ * @param  {RestClient}         client                          The REST client to make the request with
+ * @param  {Number}             [app]                           The application to search users in, defaults to the current application
```
Same comments here

nicolaasmatthijs commented on node_modules/gh-users/tests/util.js:
```diff
@@ -37,6 +37,89 @@ var assertGetMe = module.exports.assertGetMe = function(client, callback) {
+var getUsersFails = module.exports.getUsersFails = function(client, app, filter, limit, offset, code, callback) {
+    client.user.getUsers(app, filter, limit, offset, function(err, users) {
+        assert.ok(err);
+        assert.strictEqual(err.code, code);
```
Should assert `!users`

nicolaasmatthijs commented on node_modules/gh-users/tests/util.js:
```diff
@@ -37,6 +37,89 @@ var assertGetMe = module.exports.assertGetMe = function(client, callback) {
+ * @param  {User[]}             callback.users                  The users in the application
+ * @throws {AssertionError}                                     Error thrown when an assertion failed
+ */
+var getUsers = module.exports.getUsers = function(client, app, filter, limit, offset, callback) {
```
Is it worth passing in the expected users?

simong replied:
With all the randomisation and similarity matching that's going on, it's a bit tricky. Might be best to leave this to the individual tests

nicolaasmatthijs commented:
Finished review. Back to @simong for follow-up

simong commented:
Followed-up, back to @nicolaasmatthijs

nicolaasmatthijs commented:
It looks like there are a number of these in the tests:

```
Events
    Organisers
Possibly unhandled TypeError: undefined is not a function
    at null.<anonymous> (/home/travis/build/fronteerio/grasshopper/node_modules/gh-users/lib/internal/dao.js:9:4865)
    at tryCatch1 (/home/travis/build/fronteerio/grasshopper/node_modules/sequelize/node_modules/bluebird/js/main/util.js:45:21)
    at Promise$_callHandler [as _callHandler] (/home/travis/build/fronteerio/grasshopper/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:660:13)
    at Promise$_settlePromiseFromHandler [as _settlePromiseFromHandler] (/home/travis/build/fronteerio/grasshopper/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:675:18)
    at Promise$_settlePromiseAt (/home/travis/build/fronteerio/grasshopper/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:845:14)
    at Promise._settlePromiseAt (/home/travis/build/fronteerio/grasshopper/node_modules/sequelize/lib/promise.js:66:18)
    at Promise$_settlePromises [as _settlePromises] (/home/travis/build/fronteerio/grasshopper/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:988:14)
    at Promise$_rejectPromises [as _rejectPromises] (/home/travis/build/fronteerio/grasshopper/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:981:10)
    at Promise.b (domain.js:183:18)
    at Async$_consumeFunctionBuffer [as _consumeFunctionBuffer] (/home/travis/build/fronteerio/grasshopper/node_modules/sequelize/node_modules/bluebird/js/main/async.js:77:12)
    at Async$consumeFunctionBuffer (/home/travis/build/fronteerio/grasshopper/node_modules/sequelize/node_modules/bluebird/js/main/async.js:40:14)
      1) verify organisers can be added and/or removed from events
Possibly unhandled TypeError: undefined is not a function
    at null.<anonymous> (/home/travis/build/fronteerio/grasshopper/node_modules/gh-users/lib/internal/dao.js:9:4865)
    at tryCatch1 (/home/travis/build/fronteerio/grasshopper/node_modules/sequelize/node_modules/bluebird/js/main/util.js:45:21)
    at Promise$_callHandler [as _callHandler] (/home/travis/build/fronteerio/grasshopper/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:660:13)
    at Promise$_settlePromiseFromHandler [as _settlePromiseFromHandler] (/home/travis/build/fronteerio/grasshopper/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:675:18)
    at Promise$_settlePromiseAt (/home/travis/build/fronteerio/grasshopper/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:845:14)
    at Promise._settlePromiseAt (/home/travis/build/fronteerio/grasshopper/node_modules/sequelize/lib/promise.js:66:18)
    at Promise$_settlePromises [as _settlePromises] (/home/travis/build/fronteerio/grasshopper/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:988:14)
    at Promise$_rejectPromises [as _rejectPromises] (/home/travis/build/fronteerio/grasshopper/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:981:10)
    at Async$_consumeFunctionBuffer [as _consumeFunctionBuffer] (/home/travis/build/fronteerio/grasshopper/node_modules/sequelize/node_modules/bluebird/js/main/async.js:77:12)
    at Async$consumeFunctionBuffer (/home/travis/build/fronteerio/grasshopper/node_modules/sequelize/node_modules/bluebird/js/main/async.js:40:14)
    at process._tickDomainCallback (node.js:463:13)
      2) verify validation when adding and/or removing organisers from events
      3) verify authorization when adding and/or removing organisers from events
Possibly unhandled TypeError: undefined is not a function
    at null.<anonymous> (/home/travis/build/fronteerio/grasshopper/node_modules/gh-users/lib/internal/dao.js:9:4865)
    at tryCatch1 (/home/travis/build/fronteerio/grasshopper/node_modules/sequelize/node_modules/bluebird/js/main/util.js:45:21)
    at Promise$_callHandler [as _callHandler] (/home/travis/build/fronteerio/grasshopper/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:660:13)
    at Promise$_settlePromiseFromHandler [as _settlePromiseFromHandler] (/home/travis/build/fronteerio/grasshopper/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:675:18)
    at Promise$_settlePromiseAt (/home/travis/build/fronteerio/grasshopper/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:845:14)
    at Promise._settlePromiseAt (/home/travis/build/fronteerio/grasshopper/node_modules/sequelize/lib/promise.js:66:18)
    at Promise$_settlePromises [as _settlePromises] (/home/travis/build/fronteerio/grasshopper/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:988:14)
    at Promise$_rejectPromises [as _rejectPromises] (/home/travis/build/fronteerio/grasshopper/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:981:10)
    at Async$_consumeFunctionBuffer [as _consumeFunctionBuffer] (/home/travis/build/fronteerio/grasshopper/node_modules/sequelize/node_modules/bluebird/js/main/async.js:77:12)
    at Async$consumeFunctionBuffer (/home/travis/build/fronteerio/grasshopper/node_modules/sequelize/node_modules/bluebird/js/main/async.js:40:14)
    at process._tickDomainCallback (node.js:463:13)
```

The tests also seem to be ending prematurely with the following consistent error:

```
5) Series Aggregation verify serie metadata is updated when an event gets created under a serie:
     Error: timeout of 60000ms exceeded
      at null.<anonymous> (/home/travis/build/fronteerio/grasshopper/node_modules/grunt-mocha-hack/node_modules/mocha/lib/runnable.js:159:19)
      at Timer.listOnTimeout [as ontimeout] (timers.js:112:15)
```

Would you be able to have a look at this?

simong commented:
Done, the event organisers PR that got merged added a reference to the (in this PR) refactored getUsersById method. I've rebased the branch and fixed it up

nicolaasmatthijs commented:
Looks good. Merged.
